bridges/snowbridge/pallets/system/src/lib.rs

```diff
@@ -663,6 +722,40 @@ pub mod pallet {
 			let secondary_exists = Channels::<T>::contains_key(SECONDARY_GOVERNANCE_CHANNEL);
 			primary_exists && secondary_exists
 		}
+
+		pub(crate) fn do_register_token(
+			location: &Location,
+			metadata: AssetMetadata,
+			pays_fee: PaysFee<T>,
+		) -> Result<(), DispatchError> {
+			let bridge_location = Location::new(2, [GlobalConsensus(T::EthereumNetwork::get())]);
+			let mut location = location.clone();
+			location
+				.reanchor(&bridge_location, &T::UniversalLocation::get())
+				.map_err(|_| Error::<T>::LocationConversionFailed)?;
+
+			// Record the token id or fail if it has already been created
+			let token_id = TokenIdOf::convert_location(&location)
+				.ok_or(Error::<T>::LocationConversionFailed)?;
+
+			ForeignToNativeId::<T>::insert(token_id, location.clone());
+			NativeToForeignId::<T>::insert(location.clone(), token_id);
```

Comment:
`token_id` cannot be used directly as ethereum ID as it is easily derivable from the original asset Polkadot location, and someone can DoS the bridge by pre-registering IDs on Ethereum (for popular Polkadot assets).

maybe you can further hash it using some "random" salt (could be as simple as tx block number + nonce)?
then keep another map in storage `location_derived_token_id` -> `hashed_token_id`.

Reply:
So Ethereum-native asset registration on the Ethereum side uses a total different flow. Native ERC20 tokens are registered using their `H160` address:

```solidity
        // Global token registry for all tokens (both native and foreign)
        mapping(address token => TokenInfo) tokenRegistry;
        //  Additional lookup table to map foreign token id to instantiated ERC20 contract
        mapping(bytes32 foreignID => address) tokenAddressOf;
```

Then in our ethereum `Gateway.sendToken` API we use those lookup tables to discriminate between native and "foreign" ERC20 contracts: https://github.com/Snowfork/snowbridge/blob/c2af15e16b95633b2ad5939e67d4911efafaea01/contracts/src/Assets.sol#L111-L132

Reply:
> someone can DoS the bridge by pre-registering IDs on Ethereum

Btw: Register PNA cost a lot of gas and is very expensive. 12 times of the transfer cost as in in https://github.com/Snowfork/polkadot-sdk/blob/3191ea93888e97f175ef4821f908ab22fa755b41/bridges/snowbridge/primitives/core/src/outbound.rs#L461

Reply:
ok, thanks for the explanation.

for my education can you point me to the bridge sol code that registers native or foreign assets on Ethereum?
I am curious how the new ERC20 contract address is chosen/derived.

Reply:
Code:
https://github.com/Snowfork/snowbridge/blob/c2af15e16b95633b2ad5939e67d4911efafaea01/contracts/src/Assets.sol#L326-L340